server/rest/src/main/java/org/infinispan/rest/CacheInputStream.java

```diff
@@ -0,0 +1,81 @@
+package org.infinispan.rest;
+
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.infinispan.CacheStream;
+
+/**
+ * An {@link InputStream} that reads from a {@link CacheStream} of byte[] and produces a JSON output.
+ *
+ * @since 10.0
+ */
+public class CacheInputStream extends InputStream {
+   private enum State {BEGIN, START_ITEM, ITEM, END_ITEM, SEPARATOR, END, EOF}
+
+   private static final char STREAM_OPEN_CHAR = '[';
+   private static final char DATA_ENCLOSING_CHAR = '\"';
+   private static final char SEPARATOR = ',';
+   private static final char STREAM_CLOSE_CHAR = ']';
+
+   private final Iterator<?> iterator;
+   private final Stream<?> stream;
+   private final int batchSize;
+
+   private byte[] currentEntry;
+   private int cursor = 0;
+
+   private State state = State.BEGIN;
+
+   public CacheInputStream(CacheStream<?> stream, int batchSize, boolean parallel) {
+      CacheStream<?> cacheStream = parallel ? stream.parallel() : stream;
```

Comment:
I personally don't think it is worth doing this in parallel most likely. Just a single threaded per node one should be sufficient for REST.F